### ISSUE TITLE
Questions!

### ISSUE THREAD

**Issue by jeremyworger** [anakrusis/paper-sound-lua-protos#1]
Hey there, 
Does the program you've written take inputs of different resolution?  
Do you print off an image on paper and then load it in?  
I see the audio_o folder, sounds so cool!

I'd thought I'd share a similar project I've come across: https://github.com/SamusAranX/Go-MusicImage  
But I think this is lossless opposed to lossy.  I think im after lossy...if that means different resolutions or crappier images of the original image can still create sound.  
Is the format a circular image like this one? 
Also, is there a license for your project if I'd like to try it out?  
Is it possible to write a how-to for people like myself who have never used Lua? 
Cheers,

Jeremy

**Comment by anakrusis** [comment]
Hi! Thank you for checking out the project! 😀there is no license, It is all public domain and free to use however you like!

This repository contains multiple image->sound and sound->image programs, the older one being wav2img/img2wav, which is a rectangular sound format, and sounddisc, which is a circular format. And wow! MusicImage looks amazingly similar, yes! : )

Wav2img/img2wav were designed to be printed out on a sheet of paper and scanned. That is what is found in the audio_o folder.
Sounddisc was designed to be sent as an attachment to friends on Discord or another messaging platform of choice. I've tried printing out soundisc images and scanning them and they don't fare very well on paper!! Instead it has different cool lossy properties due to the pixel artifacts, which get stronger as the playhead travels further inwards.

The most recent version of sounddisc on this repository can import and read audio off any size square image, but the older version in releases (0.1) only supports a fixed resolution 2048x2048
In order to export any size image, certain constants can be tweaked within the code. The constants RENDER_W, RENDER_H determine the size of the canvas and thus the image exported. 
The constants RADIUS_START, RADIUS_END, and STRIP_BREADTH are used for tweaking the size of the disc as rendered onto the image. The function setImageConstants takes in a canvas' width and height and modifies the above three in proportion to them. Maybe in the future these could all be pulled out into a config file for easier modification, or within the GUI interface itself.

I hope it is fun to try out! If you have any more questions feel free to ask!
-Amimi

**Comment by jeremydeanw** [comment]
Awesome!  
I tried getting it to run, but maybe I'm just an idiot....I downloaded Lua (windows x64) and dragged and dropped the gui.lua from the sounddisc folder onto the exe and nothing happens, just a quick window flash? 
Oh yes I see the RENDER_W + RENDER_H in the main.lua now

**Comment by jeremyworger** [comment]
I got it working! 

A question I have is, if I wanted the max resolution output for the PNG, would I need to change bits per sample ?  
I figure if the resolution is at max, the PNG output should be a bigger file size....then if it was printed it would be able to start from a high amount of information and then the information that is lost would be of a lesser amount? 
Print it off, then lose the info?  
Is there an ideal setting for say, 1-2 minutes worth of audio? 
Resolution and sound quality wise, to get the same audio back as you put in if you just load the same image back in on the computer?  But only lose quality when printing.

**Comment by jeremydeanw** [comment]
Hi again, If I change RADIUS_START to 1, and RADIUS_END to 1000, that seems to make it start from the center, but im not sure how to remove the centerpiece.....I tried removing everything after "-- The sound data is finished drawing! now it is time to put the center stamp on it" ,  but I got a syntax error when trying to start it.  (I dont really know what im doing)  
Im assuming the different pitched oscillating noises that start at the beginning and at the end of the track have to do with the RPM always being the same? Is there some way to remedy this?

**Comment by anakrusis** [comment]
Hi, I apologize for not responding earlier! 
You don't need to change bits per sample to change the resolution of the PNG, bits per sample is for .wav export, it determines if it is an 8-bit or a 16-bit wav file
The default settings support up to 1 minute of audio, and lowering the RPM in addition to increasing the image size probably would allow for more without sacrificing quality, but I haven't tried it yet! One of the challenges with this is the line-drawing function doesn't interpolate, so it relies on itself overlapping itself to create a smooth spiral strip of sound without gaps.
(Another idea I was thinking about was, using each RGB channel to encode seperate parts of audio which could triple the length of audio supported!)

If you are planning on printing it out on paper, there must be gaps between each strip, or else the rows will blend into one another and almost completely annhilate the sound!
Additionally, with my older, orthogonal img2wav program, there is a averaging function which acts as error correction for the grain of the paper, another factor which almost completely destroys the sound! This is a little bit harder to do on a spiral because the points to average must be along a tiny piece of a line segment from the center to the playhead, I just never implemented this on the disc playback

You are correct about adaptive RPM, the current system preserves angular speed like a real record player does, whereas an ideal system would preserve linear speed and move slower along the edge and faster in the middle to compensate. This has to be done both when generating the image and playing it back (There can be an inverse relation between the distance to the center and the rpm, I believe)

Also yes! audio export would be a good feature in the future, it will be very straightforward because the soundData is fully loaded when the play button is pressed!
I hope this helps I'm sorry I took so long to respond!

**Comment by jeremydeanw** [comment]
No worries at all, thanks for getting back to me....I had a hard time figuring it out, barely know what im doing. I had never tried Lua before.  Thanks for the explanations.
An inverse relation between the distance to the center sounds like the perfect approach.  
I had sortof imagined it like "the worse quality the image = the worse quality the sound" and the more data you could resolve, then the more information you'd obtain.  Somehow!  
I don't know if I quite understand the RGB idea, but that sounds great if the quality and length can be improved, I'd definitely be curious to try a version like that.  
I'm certainly not smart enough to implement it myself, so I'll look here if you do an update on it in the future.
I haven't had the chance to try it on paper yet, more just changing the image ratio in microsoft paint, then reloading it back in.

**Comment by jeremydeanw** [comment]
Heya, I'm trying this out once again, I'm forgetting a few things so sorry if this is a dumb question..
I have a small audio clip: https://sndup.net/8np9  
The image that is created is tiny in the middle: https://www.dropbox.com/s/p3e6b2hlwvl50e9/BIG%20BAND.png?dl=0
Is there a way so that the image fills the entire disc, like: https://www.dropbox.com/s/mjmocyfsfu865ho/BIG%20BAND2.png?dl=0  

Instead?
My current settings are: 


_REVOLUTIONS_PER_MINUTE = 190


-- NOT SO CONSTANTS (they get overwritten but these are default values)

SAMPLE_RATE = 48000
SAMPLES_COUNT = SAMPLE_RATE * 60
SAMPLES_PER_TICK = math.floor(SAMPLE_RATE / 60)
BITS_PER_SAMPLE = 16
REVOLUTIONS_PER_SAMPLE = REVOLUTIONS_PER_MINUTE / 60 / SAMPLE_RATE
SAMPLES_PER_REVOLUTION = 1 / REVOLUTIONS_PER_SAMPLE_

+ 


	_RENDER_H = height	
	RENDER_W = width

	CENTER_X = RENDER_W/2
	CENTER_Y = RENDER_H/2
	-- in pixels
	RADIUS_START = ( 0 ) * ( RENDER_H / 2048 )
	RADIUS_END   = ( 2000 ) * ( RENDER_H / 2048 )
	STRIP_BREADTH = ( 5 ) *   ( RENDER_H / 2048 )_

**Comment by jeremydeanw** [comment]
Sorry I guess that doesn't make much sense since the audio needs to be a certain RPM, but maybe the tiny image could be scaled to fill a larger image and still play at the same RPM?

**Comment by jeremydeanw** [comment]
Ahh, maybe is there a way to set the pixels so that they can be out of range? 
I think that might solve the issue. It would cut off your sample, corner to corner, but then it'd fit more of the sample within the image.  Not sure if that makes any sense....

**Comment by jeremyworger** [comment]
Is there a way to make it a spiral-square/rectangle format? It would optically fit better (with more information) on a piece of paper, from center to edge, the RPM wouldn't have to change either.  That way even if you got close and cut off the image, itd play what you have resolved (I think)
I have no idea how to make the radius non-circular though !

**Comment by anakrusis** [comment]
Hi,
I think the out of bounds pixels are allowed as of commit 8485bf7742b65965b5d745f6b9ccf20ed717c915 ! Or at least, the playhead can go outside the image bounds without crashing now, I believe when 2 * RADIUS_START exceeds RENDER_W or RENDER_H, it should go outside the boundaries as shown in the image.

The square/rectangular spiral is a good idea! The way I would imagine implementing it is breaking the strips into blocks of substrips like this: https://i.imgur.com/LYDFIBj.png
The lengths of the substrips decrement seperately in the two dimensions until one of them reaches zero. These length values would be kept track for both writing strips and reading them for playback.

You might have also seen already, how these prototypes first started: https://i.imgur.com/BgIpivw.png
This is also optimal for space on the paper, but it has the discontinuity, so the slightest misalignment in scanning would cause severe popping sounds. The rectangular spiral is a better solution to this!

-Amimi

**Comment by jeremydeanw** [comment]
What about a fibonacci-like substrip? https://imgur.com/a/jL4mrYO

**Comment by jeremydeanw** [comment]
Alternatively....https://imgur.com/a/tcrK1Gt (super great MS paint skills) could have all the same size of strip, read them all top to bottom/bottom to top, but still read them as a square spiral, block to block. That way when the image is scaled, all the mini-cube strips would shrink or expand, but still stay cubes. That's probably alot easier than all these weird sized blocks, because that would be hell to code, I would have to imagine.

**Comment by jeremydeanw** [comment]
I realized this doesn't make alot of sense, it would take up way too much space.  If there is a way to non invasively change the current code to a square spiral though, thatd be cool to try.  Cheers

**Comment by jeremydeanw** [comment]
Ahh! I thought of another idea!  
https://imgur.com/a/DAhfYOU
Have a single row of variable density pixels exit from the center of the image, while simultaneously have a duplicate of that row returning next to it in the opposite direction. 
That way, say if you got 3/4 of the image cut off at the top, then you could fill in the missing 1/4 row by obtaining it right next to it towards the center.  
Then just have a bunch of paired lines all exiting the center outwards. Which would also allow full rotation I think.  
It doubles the amount of space needed, but at least it wouldn't quadroople it like the mini-squares.  
I don't know what multiple lines exiting the center would look like though, pixel wise, like how to get the lines on an angle.

**Comment by jeremydeanw** [comment]
What about just interlocking two spirals together? Exactly how you have it, but have the same audio playing in the opposite direction right next to it?  
Either that or somehow merge the reversed track with the regular track, and somehow sonically separate them afterwards?  
I think this would allow cropping, because the midway point would be midway in the image itself. Since it'd be playing in both directions.

**Comment by jeremydeanw** [comment]
- Maybe some type of lossy JPEG steganography? To hide two strips on one strip?

**Comment by jeremyworger** [comment]
Sorry for opening and closing so many times. I tried to get all my questions in one place, but it ended up a bit messy.  I had this other totally different idea, but I didn't want to interfere with what you've already got going on with the current format. Thankyou for the help above.

**Comment by anakrusis** [comment]
No worries, I appreciate your feedback and advice on these little experiments, and I apologize for not responding very timely on your questions, as this has been a pretty busy week here. It has been an interesting and enjoyable discussion, and I didn't feel it was interfereing-- although admittedly I haven't updated the project in a while-- and you are always welcome to come back anytime and try out more stuff with the programs if you want!

**Comment by jeremydeanw** [comment]
Thanks :) I do want to share the idea I had (had to do with color output) but I had larger attachments I can't upload here....so my email is mrjeremydean@gmail.com if you wanted to hear it. Cheers again!

**Comment by jeremydeanw** [comment]
I'd definitely be interested in the idea you had posted for the square rectangular format above, I just wasn't sure how to implement it myself :< I assumed it has to do with changing the RADIUS function. I had another idea for color and cropping too. (sorry for the extra message).

**Comment by jeremydeanw** [comment]
Hey so im back to using commit #8485bf7 (since I couldn't figure out how to remove the centerpiece in the current version) but I'm still getting an error for the out of bounds pixel thingy, error was 105, attempt to set out-of-bounds pixels?

**Comment by jeremydeanw** [comment]
How I had imagined the square rectangle: https://imgur.com/a/FaUiFpN :)

**Comment by jeremydeanw** [comment]
But still grayscale (in case my bad drawing is confusing).

**Comment by jeremydeanw** [comment]
So im trying out the "[audio strip format fully functional]" version from July 16th '21....is there a way to make the strips so that they are flush or even overlapping?  How do I change the strip width?
When I try to crop the image and then load it back in I get the error:

"attempt to perform arimetic on global 'render_h' (a nil value)  

Maybe if there was the same track, but reversed, contained within each strip, then you could play from either side towards the center? But still left to right.

**Comment by jeremydeanw** [comment]
The only other way I can think of would have dead squares: 
https://www.dropbox.com/s/x5tqxbdy76ef1gf/Dead%20Squares.jpg?dl=0  
So when you are cropping the image none of the strips are lost.

**Comment by jeremydeanw** [comment]
What about stereo variable density strips, and each strip contains a reversed part of itself:
https://www.dropbox.com/s/1ifxev382e25mro/variable%20density%20multidirection.png?dl=0
So L channel would be regular, R channel would be reverse, on the same strip. So it could be read from either side, also allowing cropping. (and rotation, if the strips are bigger)  But would double the data required, unless there is some fancy way of coding stereo variable tracks in binary.

**Comment by jeremydeanw** [comment]
Just realized that doesn't make any sense! Because the start would always be cut off. Will delete!

**Comment by jeremydeanw** [comment]
Im checking out: http://stevehanov.ca/wavelet/  
There is an example here of selecting (or cropping in this case) a part of the image and still being able to play it. I wonder if this same concept can be applied with variable density strips, having the whole audio cut into strips by frequency going horizontally: 
https://www.dropbox.com/s/mzemjmce6ga4roa/variable%20frequency%20split.png?dl=0

**Comment by jeremydeanw** [comment]
https://github.com/WIZARDISHUNGRY/sox-slice  
Can separate a sound into frequency bands, I've been messing around with it and trying to figure out what range and how many bands would be best. (for human hearing range) A band per variable density strip I was thinking, not sure how to make that happen though, but was curious to know what you thought about this.  
The output is quite large! : https://www.dropbox.com/s/vnugy57l6svxhwj/output%20hz.png?dl=0

**Comment by jeremydeanw** [comment]
Wondering two things,  (for the July 19th Variable Density commit)  
- Is there a way to play all strips at once? Either left/right or up/down.  
- If theoretically a sound file can be separated into frequency band files, can a band be loaded per strip?

**Comment by jeremyworger** [comment]
Im wondering if something like: https://github.com/phamquiluan/jdeskew  
Could work in theory

**Comment by anakrusis** [comment]
Hi! this would probably help a lot, even just seeing how it is implemented, because it has indeed been an issue when I have gone to scan it, that it is slightly diagonally off and this makes it very difficult to read

To answer your earlier questions (sorry I missed them) the playback is only sequential from one strip to the next as of right now. I was considering splitting the sound into a bass band and a treble band, and compressing the bass band, which would theoretically work, but also the playback would have to adjust to play the two strips simultaneously... it is all things that can be done later probably!

**Comment by jeremydeanw** [comment]
Yeah! Sorry im not much of a githuber, so I opened and closed it a bunch haha. 
I couldn't find anything Lua related....I think for this one, it's actually automatic? Python though.  
I've seen deskewing before, but usually the user has to select the corners and stuff.  
I was thinking at least like, 10 bands simultaneously?  I'm thinking that a deskewer could recognize the "document" angle for variable density because it's already in rectangular strips....I guess it would be trial and error.  
I was sortof imagining it to looking something like the PDF417 barcode?  (but without alignment bars!) 
Also I think that would resolve cropping issues, it'd allow you to crop inside the image, mind you losing the start and beginning of the track.....but I personally wouldn't care much about that, I think it'd be cool!

**Comment by jeremydeanw** [comment]
Ok....what about: https://www.dropbox.com/s/501dolqtqxbqnim/Idea.png?dl=0  
https://www.dropbox.com/s/coafgp5dhfnlwsz/spectro.png?dl=0  (but variable density)  
Where VD strips are all heading towards the center, and are all read simultaneously from the outside inwards at the same time?
(it wouldn't actually be circular like this, all the strips would be touching/coming from the outside edge of the image) 
So it wouldn't necessarily matter where the center lies, just that you'd be cutting off some of the beginning of the track when you crop too much.  Speed being the main input option.
Since the strips would have to be like a pie, the information in the center would be somewhat lost due to low resolution, so the end of the track would start to become muffled as reading gets closer to the center of the image.
That way input angle of the image wouldn't matter at all. Basically reading it at every possible pixel angle. 
Im not sure if that makes sense though. Just what I've been thinking about.

**Comment by jeremydeanw** [comment]
Oh nooo nevermind, that doesn't make sense, if the center is off even slightly then it will be all messed up.  Unless the center was in multiple locations.

**Comment by jeremydeanw** [comment]
Perhaps a solution? 
https://github.com/xiaoyu258/GeoProj  
With a license too!
I don't know much about neural stuff, but there is a pre-trained model.

**Comment by jeremydeanw** [comment]
https://github.com/phamquiluan/jdeskew 
Has a license now, so probably easier to work with compared to the other one I just posted. (I assume)

**Comment by jeremydeanw** [comment]
I have it like: https://www.dropbox.com/s/kbeq8pno79t3pro/TEST.png?dl=0  
But yeah the strips would need to all play from left to right. Not sure how to implement that.  
I'm also not sure how to decrease the spaces between the strips either, but I imagine this as it is should be able to be deskewed if printed using the jdeskew library above. I'm going to give that a go, and ill let you know if it works.

**Comment by jeremydeanw** [comment]
I tested with jdeskew,  
Input: https://www.dropbox.com/s/qmt5cads6l3l93r/input.png?dl=0  
Output:  https://www.dropbox.com/s/dmfognipyy61507/output.png?dl=0  

So although it works fine (other than that there seems to be a limit of input image size) it won't really help because even with rotation correction, there still needs to be perspective correction. The strips are only straight in some places, but this is because of how I took the picture casually.  
The best bet is indeed the other one.

**Comment by jeremydeanw** [comment]
What about variable area strips, but each strip is represented by a single squiggly line (so no borders). Then just have a bunch of squiggly lines left to right, each representing a frequency. With a small bit of space between each line.  
Then if the image is slightly cropped or rotated or skewed, perform line extraction on every line within the image (reading the lines like vectors more or less), and then read all lines simultaneously.  ?

**Comment by jeremydeanw** [comment]
Maybe there wouldn't have to be any extraction at all.  Could it be done via nearest neighbor, in a way where the lines are read by following dark pixels no matter which direction the line is going?
https://cloud.githubusercontent.com/assets/2564583/20510932/dd6f7f7c-b028-11e6-94df-dd1693214b3a.png

**Comment by jeremydeanw** [comment]
I have another idea! 
What if simply every pixels shade in a row represent a variable area height? Height between 0-255 in a "space" that isn't physically on the paper.  
Like: https://www.dropbox.com/s/xl1oh5xi1faulaq/height.png?dl=0  
How this helps with rotation and skewing I don't know....but I feel like there is something here

**Comment by jeremydeanw** [comment]
With variable density, would it be possible to just have one huge compacted track as the image, like lines from the top to the bottom of the image?  Because then it could be read and cropped at any angle since the whole image would be made from the same track.   
I was thinking that if the lines are interpreted/averaged as "dark" or "light" when they are being read left to right or whatever, then you could encode extra information within those lines by just changing the values of the pixels within the lines slightly.  Like a line made up of not just black, but black and darker grays as well....etc. There might even be a way of splitting the track in two and having it go through itself? Just a thought

**Comment by jeremydeanw** [comment]
Like: https://www.dropbox.com/s/7z3zcrdez64y1el/markers.png?dl=0 
I don't know what or where these markers would be, but yeah, if the track is able to be read as just "light" or "dark" then I think it might be possible to layer variable density 1D tracks in each other...I think as long as the markers are somewhat near the center as opposed to the outsides.

**Comment by jeremydeanw** [comment]
More or less what I was wondering is if a single line was made up of 3 (probably at the most) different brightnesses, could different brightness passes be made to the same variable density strip? Because if it's 1D (but multiple layers), then that really solves all the problems, and excludes the need for corrections. Not sure how many seconds you'd be able to get in one place though

**Comment by jeremydeanw** [comment]
Actually that doesn't make much sense since time is still left to right. But maybe there is a way to do different passes of brightness if the image is split into some kind of 1D variable density looking frequency bands

**Comment by jeremydeanw** [comment]
For variable area, I was thinking line following: https://www.dropbox.com/s/h6uwajfb22axrqt/Untitled.png?dl=0 would work at any rotation, even with skewing.  Each squiggle being a frequency all played at once. But I have no idea how lines would be detected and followed :-I

**Comment by jeremydeanw** [comment]
I think squiggly variable area lines could work with only nearest neighbour, no edge detection, opencv...etc. Everytime a black pixel is found starting on one side, you could follow the black pixels based on the non-black pixels around them. Then once you reach the tips of the squiggle, always have it look for more black pixels around from its location and head back....etc  
Thoughts?

**Comment by jeremydeanw** [comment]
Also what happens if you read a horizontal variable area vertically? Like if this image were on an angle https://www.dropbox.com/s/1fw7p6459o7pn2s/sLJgkeEo.png?dl=0  If each line of pixels contains small changes, could each line be converted to bad sounding-sound, then could the difference in sound be pieced together afterwards?
(probably the weirdest idea yet)

**Comment by jeremydeanw** [comment]
Actually since you already have the variable density strips spaced out in the current version, would it be possible to just surround the strips with a thin layer of black pixels, then read the image starting from the left side and say to follow each strip based on the difference between black and white pixels? White being the space between the strips. Basically just following the black pixels and calculating what lies beneath the black pixels? By nearest neighbour line following (up and over) and each strip being a frequency or as you said bass/treble or something similar?  
Basically to have something surrounding the strips that is understood as different from the strips themselves and that would indicate the beginning of a new strip to be read

**Comment by jeremydeanw** [comment]
I found something interesting: 

https://www.dropbox.com/s/wstvq3l14omupjl/1.png?dl=0
https://www.dropbox.com/s/paf06j7mrbhxyn3/2.png?dl=0

From: https://crypto.stanford.edu/RealWorldCrypto/slides/Garay.pdf

It mentions "BFS" which I think is breadth first search?  
It seems that they are able to read the image at all angles and crops, I think 
"can be read regardless of orientation or skew ― “orientation robustness""  

It actually looks similar to variable density in a way. Maybe this type of BFS concept can be applied in some way?

**Comment by jeremydeanw** [comment]
Actually what about encoding a number per "block of pixels" of a certain size within the image? If it were possible to break down the sound into pixel blocks of oscillations, but also labeling each block with a number that could somehow be recognized visually. So it could be read block by block (in a spiral), regardless of the captured image orientation, like 1, 2, 3, 4

**Comment by jeremydeanw** [comment]
Woah I just had an idea, what if you just broke down the variable density soundtrack into a grid? So https://i.stack.imgur.com/RNoXW.jpg (obviously no numbers) but multiple of these going around in a spiral, and each one containing a number (not sure of this part) so you know the order after capturing. Because each mini 1d soundtrack could be read at any angle because they are 1D

**Comment by jeremydeanw** [comment]
Like: https://www.dropbox.com/s/j61hqpn1ug1z9qk/multi%201D.png?dl=0

**Comment by jeremydeanw** [comment]
Yeah, is it possible to turn the strips into a grid of colored mini strips? I wouldn't know how to begin haha. Just wondering if that would work
